moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp

```diff
@@ -391,53 +403,91 @@ void planning_scene_monitor::CurrentStateMonitor::jointStateCallback(const senso
           robot_state_.setJointPositions(jm, &b.max_position_);
       }
     }
+  }
 
-    // read root transform, if needed
-    if (tf_ && (robot_model_->getRootJoint()->getType() == robot_model::JointModel::PLANAR ||
-                robot_model_->getRootJoint()->getType() == robot_model::JointModel::FLOATING))
+  // callbacks, if needed
+  if (update)
+    for (std::size_t i = 0; i < update_callbacks_.size(); ++i)
+      update_callbacks_[i](joint_state);
+
+  // notify waitForCurrentState() *after* potential update callbacks
+  state_update_condition_.notify_all();
+}
+
+void planning_scene_monitor::CurrentStateMonitor::tfCallback()
+{
+  bool update = false;
+  bool changes = false;
+
+  // read multi-dof joint states from TF, if needed
+  const std::vector<const moveit::core::JointModel*>& multi_dof_joints = robot_model_->getMultiDOFJointModels();
+
+  {
+    for (const moveit::core::JointModel* joint : multi_dof_joints)
     {
-      const std::string& child_frame = robot_model_->getRootLink()->getName();
-      const std::string& parent_frame = robot_model_->getModelFrame();
+      const std::string& child_frame = joint->getChildLinkModel()->getName();
+      const std::string& parent_frame =
+          joint->getParentLinkModel() ? joint->getParentLinkModel()->getName() : robot_model_->getModelFrame();
 
       std::string err;
       ros::Time tm;
-      tf::StampedTransform transf;
-      bool ok = false;
-      if (tf_->getLatestCommonTime(parent_frame, child_frame, tm, &err) == tf::NO_ERROR)
+      if (!tf_->getLatestCommonTime(parent_frame, child_frame, tm, &err) == tf::NO_ERROR)
       {
+        ROS_DEBUG_STREAM_THROTTLE(1, "Unable to update multi-dof joint '"
+                                         << joint->getName() << "': tf has no common time between '"
+                                         << parent_frame.c_str() << "' and '" << child_frame.c_str() << "': " << err);
+        continue;
+      }
+
+      const std::vector<std::string>& joint_vars = joint->getVariableNames();
+      assert(joint_vars.size() > 0);
+      if (joint_time_[joint_vars[0]] != tm)
```

Comment:
I think this should use the joint name instead of joint's variable name(s). This also resolves ambiguity, which one to use - why did you choose `joint_vars[0]`?

Reply:
Is it safe to read-access `joint_time_` without locking? A writing thread may change the map meanwhile...

Reply:
of course you are right about locking.. fixed.

Reply:
Also I adjusted the map as you proposed. Nice idea 👍 
The previous implementation was correct though.
joint_vars[0] represents the first dof of the joint *including the joint name prefix*, e.g. `joint_name/trans_x`. I chose the first one as a representation for all dof of this joint.
It's much better to just store the time per joint instead and apparently it doesn't break anything else in MoveIt.